nx/lib/nx/lin_alg.ex

```diff
@@ -1759,4 +1759,72 @@ defmodule Nx.LinAlg do
     |> Nx.take_diagonal(offset: offset)
     |> Nx.product(axes: [rank - 2])
   end
+
+  @doc """
+  Return matrix rank of input M × N matrix using Singular Value Decomposition method.
+
+  Approximate the number of linearly independent rows by calculating the number
+  of singular values greater than `eps * max(singular values) * max(M, N)`.
+
+  This also appears in Numerical recipes in the discussion of SVD solutions for
+  linear least squares [1].
+
+  [1] W. H. Press, S. A. Teukolsky, W. T. Vetterling and B. P. Flannery,
+  “Numerical Recipes (3rd edition)”, Cambridge University Press, 2007, page 795.
+
+  ## Options
+    * `:eps` - Rounding error threshold used to assume values as 0. Defaults to `1.0e-10`
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 2], [3, 4]]))
+      #Nx.Tensor<
+        u64
+        2
+      >
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 1, 1, 1], [1, 1, 1, 1], [1, 2, 3, 4]]))
+      #Nx.Tensor<
+        u64
+        2
+      >
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([[1, 1, 1], [2, 2, 2], [8, 9, 10], [-2, 1, 5]]))
+      #Nx.Tensor<
+        u64
+        3
+      >
+
+  ## Error cases
+
+      iex> Nx.LinAlg.matrix_rank(Nx.tensor([1, 2, 3]))
+      ** (ArgumentError) tensor must have at least rank 2, got rank 1 with shape {3}
+  """
+  @doc from_backend: false
+  def matrix_rank(a, opts \\ []) do
```

Comment:
```suggestion
  defn matrix_rank(a, opts \\ []) do
```

Can we use defn here?

Reply:
@polvalente 
Since we use function `svd/1` in function `matrix_rank`, we cannot use `defn` for function `matrix_rank`.
https://github.com/elixir-nx/nx/actions/runs/3982399082/jobs/6826831892

Reply:
What happens if you use the fully qualified Nx.LinAlg.svd instead of calling it locally?

Reply:
Thanks for the advice, I did successfully call `Nx.LinAlg.svd` in defn `matrix_rank`.
I also fixed `ArgumentError` in defn instead of `Nx.Shape.matrix_rank` because `Nx.Shape.matrix_rank` cannot be called in defn.  [3fafb9b](https://github.com/elixir-nx/nx/pull/1071/commits/3fafb9b657244e868d594bc4934c3d6f11bfd55d)